### PR TITLE
fix(grpo,rloo): apply generation_config override in use_transformers_paged path

### LINKED ISSUE #5783: GRPOTrainer silently uses near-greedy decoding when temperature=1.0 (transformers >= 4.50 + Qwen2.5)
## Summary

`GRPOTrainer` silently runs near-greedy decoding when `GRPOConfig.temperature=1.0` (the default) is used with any Qwen2.5-family model since `transformers >= 4.50.0`. This collapses GRPO rollouts to near-duplicates, making the advantage signal degenerate (`std(R) ≈ 0`) and silently breaking the algorithm's exploration mechanism. Training runs to completion with no error.

## Reproduction

Same SFT-trained LoRA adapter, same prompt, `temperature=1.0`:

* **Direct `model.generate()`** (32 samples) → 7 distinct outputs in 13 parses ✓ diverse
* **TRL GRPOTrainer** (8 rollouts, `num_generations=8`) → 4 of 8 completions have **identical md5 hashes** ❌

Minimal repro (Qwen2.5-VL-3B):
```python
import torch
from transformers import GenerationConfig

# === BUGGY (matches TRL behavior) ===
gen_config = GenerationConfig(
    temperature=1.0, top_p=1.0, top_k=None,
    repetition_penalty=1.0, do_sample=True, max_new_tokens=512,
)
out_buggy = model.generate(input_ids=batch_8_duplicated_rows, generation_config=gen_config)
# → 4 unique md5s / 8 rows

# === FIXED ===
out_fixed = model.generate(
    input_ids=batch_8_duplicated_rows, generation_config=gen_config,
    use_model_defaults=False,  # ← the fix
)
# → 8 unique md5s / 8 rows
```

Buggy path prints this warning (easily missed):
```
`generation_config` default values have been modified to match model-specific defaults:
{'max_length': 128000, 'temperature': 1e-06, 'repetition_penalty': 1.05, 'bos_token_id': 151643}.
```

## Root cause

`transformers.generation.utils.GenerationMixin._prepare_generation_config` (introduced in transformers 4.50.0) silently overrides user-set fields in `GenerationConfig` that equal the **global defaults** with the model's bundled `generation_config.json` values, unless the caller passes `use_model_defaults=False` to `model.generate(...)`.

Qwen2.5-VL-3B-Instruct's bundled `generation_config.json` ships with:
```json
{
  "temperature": 1e-06,
  "top_p": 0.001,
  "top_k": 1,
  "repetition_penalty": 1.05,
  "transformers_version": "4.50.dev0"
}
```

`GRPOTrainer.__init__` (around `grpo_trainer.py:564-579`) builds `self.generation_config` directly from `GRPOConfig` defaults (1.0, 1.0, None, 1.0) — every field happens to be a global default — and `_generate_single_turn` (around `grpo_trainer.py:1292`) calls `unwrapped_model.generate(generation_config=self.generation_config, ...)` **without** passing `use_model_defaults=False`. All four sampling fields silently revert to Qwen's near-greedy values inside `model.generate()`.

Effect on GRPO: `num_generations` rollouts per prompt become near-duplicates → `std(R) ≈ 0` in the advantage formula → effectively no gradient signal for exploration.

## Why this matters

We discovered this while investigating "Red Agent mode collapse" in an adversarial training setup. After ~6 weeks of trying every 2025-recommended exploration remedy — DAPO clip-higher (arXiv:2503.14476), GAPO group-frequency reward (arXiv:2511.12596), Active Attacks defender refresh (arXiv:2509.21947), NGRPO advantage calibration (arXiv:2509.18851) — none of them worked because the rollouts were never actually diverse to begin with. After applying the fix below, within-group diversity is restored on the same training setup.

Any HuggingFace user running GRPO on a Qwen-family model with default `GRPOConfig.temperature=1.0` since `transformers 4.50.0` (released ~2025-01) is silently affected.

## Proposed fix

Pass `use_model_defaults=False` in `_generate_single_turn` (and `_generate_single_turn_vllm` doesn't need it — vLLM bypasses this code path):

```python
prompt_completion_ids = unwrapped_model.generate(
    **generate_inputs,
    generation_config=self.generation_config,
    disable_compile=True,
    use_model_defaults=False,   # ← prevent silent override
)
```

This documents intent: TRL has carefully constructed `self.generation_config` from user-provided `GRPOConfig` and the user expects those values to be used as-is, not silently merged with the model's bundled config.

## Workaround until upstream fix

Subclass `GRPOTrainer` (or any of its descendants) and bump default-equal sampling fields by 1 ulp in `__init__`:

```python
def __init__(self, *args, **kwargs):
    super().__init__(*args, **kwargs)
    gc = self.generation_config
    if float(getattr(gc, "temperature", 1.0)) == 1.0:
        gc.temperature = 1.0000001          # bit-equivalent sampling
    if getattr(gc, "top_p", None) == 1.0:
        gc.top_p = 0.9999999
    if getattr(gc, "top_k", None) is None:
        gc.top_k = 0                         # 0 = disabled (same as None)
    if float(getattr(gc, "repetition_penalty", 1.0)) == 1.0:
        gc.repetition_penalty = 1.0000001
```

Because none of these values equal the global defaults anymore, `_prepare_generation_config` leaves them alone. Sampling distribution unchanged at the floating-point level.

## Affected versions

- TRL: 0.18 through at least 0.24.0 (any version calling `model.generate(generation_config=...)` in the generation path)
- transformers: ≥ 4.50.0
- Confirmed on: `Qwen/Qwen2.5-VL-3B-Instruct`, `Qwen/Qwen2.5-VL-7B-Instruct-unsloth-bnb-4bit`
- Likely affects: Qwen2.5-Instruct text-only and other recent Qwen variants whose bundled `generation_config.json` declares `transformers_version >= 4.50.0` and ships with non-default sampling values

## Environment

- `trl==0.24.0`
- `transformers==4.57.6`
- `vllm==0.15.1` (TRL's vLLM rollout path is **not** affected — it bypasses HF GenerationConfig and uses `SamplingParams` directly)
- Qwen2.5-VL-3B-Instruct (Unsloth bf16 + LoRA, also unsloth-bnb-4bit)

### PR DESCRIPTION
**Pull request by Sumu004**
# What does this PR do?

The `use_transformers_paged` generation path in both `GRPOTrainer` and `RLOOTrainer` was missing the `generation_kwargs` override that the regular path already applies via `_override_model_generation_config`.

On transformers < 5.0, model-specific `generation_config` values (e.g. Qwen2.5 ships with `temperature=0.7`) are silently merged on top of training kwargs during `model.generate()` — see transformers#42762. The regular generation path was already fixed (after the v0.24.0 tag) by passing `generation_kwargs=self.generation_kwargs` to `unwrap_model_for_generation`. The `use_transformers_paged` branch called `unwrap_model_for_generation` without `generation_kwargs`, leaving the same silent bug active: rollouts collapse to near-duplicates, advantage variance → 0, and GRPO/RLOO training silently fails with no error or warning.

Fix: pass `generation_kwargs=self.generation_kwargs` in the `use_transformers_paged` branch of both trainers, matching the comment and pattern already on the regular path. Also adds unit tests for `_override_model_generation_config` (none existed before).

Fixes #5783

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. — #5783
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@lewtun @kashif — this touches the GRPO and RLOO generation paths. Happy to make any adjustments.

---

> [!NOTE]
> **Low Risk**
> Small, targeted parity fix in rollout generation plus unit tests; no new APIs or auth/data paths.
> 
> **Overview**
> **Paged Transformers rollout generation** in `GRPOTrainer` and `RLOOTrainer` now passes `generation_kwargs=self.generation_kwargs` into `unwrap_model_for_generation`, matching the non-paged path. That applies the existing `_override_model_generation_config` workaround for [transformers#42762](https://github.com/huggingface/transformers/issues/42762) on transformers &lt; 5.0, so model-shipped defaults (e.g. low temperature on Qwen2.5) no longer override training sampling during `generate_batch`.
> 
> **Tests** add `TestOverrideModelGenerationConfig` in `tests/test_model_utils.py` covering temporary override, restore after context, no-op when kwargs are `None`, and no-op on transformers ≥ 5.0.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 209ac8cb85ecda1e2981519dbf7407bd7dbc05df. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
thanks a lot for the contribution, but this branch will be removed soon (no usage, inefficient, and lacking testing and support). Consequently, I'd try not to modify nor extending it in the meantime
